src/backend/joanie/core/flows/order.py

```diff
@@ -25,92 +24,166 @@ def _set_order_state(self, value):
-        payment_backend = get_payment_backend()
-        if credit_card_id:
-            try:
-                credit_card = CreditCard.objects.get(
-                    owner=self.instance.owner, id=credit_card_id
-                )
-                return payment_backend.create_one_click_payment(
-                    order=self.instance,
-                    billing_address=billing_address,
-                    credit_card_token=credit_card.token,
-                )
-            except (CreditCard.DoesNotExist, NotImplementedError):
-                pass
-        payment_info = payment_backend.create_payment(
-            order=self.instance, billing_address=billing_address
-        )
-
-        return payment_info
+    def to_sign_and_to_save_payment_method(self):
+        """
+        Transition order to to_sign_and_to_save_payment_method state.
+        """
 
     @state.transition(
         source=[
-            enums.ORDER_STATE_DRAFT,
-            enums.ORDER_STATE_SUBMITTED,
+            enums.ORDER_STATE_ASSIGNED,
+            enums.ORDER_STATE_TO_SIGN_AND_TO_SAVE_PAYMENT_METHOD,
         ],
-        target=enums.ORDER_STATE_VALIDATED,
-        conditions=[_can_be_state_validated],
+        target=enums.ORDER_STATE_TO_SAVE_PAYMENT_METHOD,
+        conditions=[_can_be_state_to_save_payment_method],
     )
-    def validate(self):
+    def to_save_payment_method(self):
         """
-        Transition order to validated state.
+        Transition order to to_save_payment_method state.
         """
 
     @state.transition(
-        source=fsm.State.ANY,
-        target=enums.ORDER_STATE_CANCELED,
+        source=[
+            enums.ORDER_STATE_ASSIGNED,
+            enums.ORDER_STATE_TO_SIGN_AND_TO_SAVE_PAYMENT_METHOD,
+        ],
+        target=enums.ORDER_STATE_TO_SIGN,
+        conditions=[_can_be_state_to_sign],
     )
-    def cancel(self):
+    def to_sign(self):
         """
-        Mark order instance as "canceled".
+        Transition order to to_sign state.
         """
 
+    def _can_be_state_pending(self):
+        """
+        An order state can be set to pending if the order is not free
+        and has a payment method and no contract to sign.
+        """
+        return (
+            self.instance.is_free or self.instance.has_payment_method
+        ) and not self.instance.has_unsigned_contract
+
     @state.transition(
-        source=[enums.ORDER_STATE_SUBMITTED, enums.ORDER_STATE_VALIDATED],
+        source=[
+            enums.ORDER_STATE_ASSIGNED,
+            enums.ORDER_STATE_TO_SIGN_AND_TO_SAVE_PAYMENT_METHOD,
+            enums.ORDER_STATE_TO_SAVE_PAYMENT_METHOD,
+            enums.ORDER_STATE_TO_SIGN,
+        ],
         target=enums.ORDER_STATE_PENDING,
+        conditions=[_can_be_state_pending],
+    )
+    def pending(self):
+        """
+        Transition order to pending state.
+        """
+
+    def update(self):
+        """
+        Update the order state.
+        """
```

Comment:
What do you think to add transitions to `pending_payment`, `failed_payment` and `no_payment`?

In this way, we would be able to just call `order.update` each time we want to update its state

Of course, it could be a TODO

Reply:
I can give it a try.

Reply:
I added a TODO